lua/lib/lfospec.lua

```diff
@@ -0,0 +1,449 @@
+  i.period = period == nil and 4 or period
+  i.reset_target = 'floor'
+  i.baseline = 'min'
+  i.offset = 0
+  i.ppqn = 96
+  i.controlspec = {
+    warp = 'linear',
+    step = 0.01,
+    units = '',
+    quantum = 0.01,
+    wrap = false,
+    formatter = nil
+  }
+  i.action = fn == nil and (function(scaled, raw) end) or fn
+  return i
+end
+
+-- PARAMETERS UI /
+local function lfo_params_visibility(state, i)
+  params[state](params, "lfo_baseline_"..i)
+  params[state](params, "lfo_offset_"..i)
+  params[state](params, "lfo_depth_"..i)
+  params[state](params, "lfo_scaled_"..i)
+  params[state](params, "lfo_raw_"..i)
+  params[state](params, "lfo_mode_"..i)
+  if state == "show" then
+    if params:get("lfo_mode_"..i) == 1 then
+      params:hide("lfo_free_"..i)
+      params:show("lfo_clocked_"..i)
+    elseif params:get("lfo_mode_"..i) == 2 then
+      params:hide("lfo_clocked_"..i)
+      params:show("lfo_free_"..i)
+    end
+  else
+    params:hide("lfo_clocked_"..i)
+    params:hide("lfo_free_"..i)
+  end
+  params[state](params, "lfo_shape_"..i)
+  params[state](params, "lfo_min_"..i)
+  params[state](params, "lfo_max_"..i)
+  params[state](params, "lfo_reset_"..i)
+  params[state](params, "lfo_reset_target_"..i)
+  _menu.rebuild_params()
+end
+
+local function build_lfo_spec(lfo,i,bound)
+  params:add{
+    type = 'control',
+    id = 'lfo_'..bound..'_'..i,
+    name = 'lfo '..bound,
+    controlspec = controlspec.new(
+      lfo.min,
+      lfo.max,
+      lfo.controlspec.warp,
+      lfo.controlspec.step,
+      bound == 'min' and lfo.min or lfo.max,
+      lfo.controlspec.units,
+      lfo.controlspec.quantum,
+      lfo.controlspec.wrap
+    ),
+    formatter = lfo.controlspec.formatter
+  }
+  params:set_action('lfo_'..bound..'_'..i, function(x)
+    lfo:set(bound,x)
+  end)
+end
+
+local function lfo_bang(i)
+  local function lb(prm)
+    params:lookup_param("lfo_"..prm.."_"..i):bang()
+  end
+  lb('depth')
+  lb('min')
+  lb('max')
+  lb('baseline')
+  lb('offset')
+  lb('mode')
+  lb('clocked')
+  lb('free')
+  lb('shape')
+  lb('reset')
+  lb('reset_target')
+end
+
+-- / PARAMETERS UI
+
+-- SCRIPTING /
+
+local function process_lfo(id)
+  local _lfo = id
+  local phase
+
+  _lfo.phase_counter = _lfo.phase_counter + (1/_lfo.ppqn)
+  if _lfo.mode == "clocked" then
+    phase = _lfo.phase_counter / _lfo.period
+  else
+    phase = _lfo.phase_counter * clock.get_beat_sec() / (1/_lfo.period)
+  end
+  phase = phase % 1
+
+  if _lfo.enabled == 1 then
+
+    local current_val = (math.sin(2*math.pi*phase) + 1)/2
+    if _lfo.shape == "saw" then
+      current_val = phase < 0.5 and phase/0.5 or 1-(phase-0.5)/(0.5)
+    end
+
+    local min = _lfo.min
+    local max = _lfo.max
+
+    if min > max then
+      local old_min = min
+      local old_max = max
+      min = old_max
+      max = old_min
+    end
+
+    if min == -inf or max == -inf then
+      min = min == -inf and -2^1023.999999999 or min
```

Comment:
use `math.huge` and `-math.huge` instead for legibility

also there is redundancy by having the `if` above this line (just remove the if check)

Reply:
- in this per-tick `process_lfo` i see many calulations involving these `min`/`max` deriverd from `_lfo.min` etc., but that does not seem to be updated here. these intermediate calculations should really just be performed whenever the lfo bounds change....

Reply:
- whoa! these `2^x` calls with constants don't work like they would in a compiled environemnt - it's really gonna recalculate these on each tick with 64b precision. just use constants

Reply:
omg, yes! changed all the min/max/depth/percentage calculations to static functions which only execute when changes actually occur.